dot/rpc/modules/chain_integration_test.go

```diff
@@ -0,0 +1,423 @@
+	req := ChainBlockNumberRequest{nums}
+
+	err := svc.GetBlockHash(nil, &req, &res)
+	require.Nil(t, err)
+
+	expected0, err := state.Block.GetBlockByNumber(big.NewInt(0))
+	require.NoError(t, err)
+	expected1, err := state.Block.GetBlockByNumber(big.NewInt(1))
+	require.NoError(t, err)
+	expected := []string{expected0.Header.Hash().String(), expected1.Header.Hash().String()}
+
+	require.Equal(t, expected, res)
+}
+
+func TestChainGetFinalizedHead(t *testing.T) {
+	state := newTestStateService(t)
+	svc := NewChainModule(state.Block)
+	_, _, genesisHeader := genesis.NewTestGenesisWithTrieAndHeader(t)
+	var res ChainHashResponse
+	err := svc.GetFinalizedHead(nil, &EmptyRequest{}, &res)
+	require.NoError(t, err)
+
+	expected := genesisHeader.Hash()
+	require.Equal(t, common.BytesToHex(expected[:]), res)
+}
+
+func TestChainGetFinalizedHeadByRound(t *testing.T) {
+	state := newTestStateService(t)
+	svc := NewChainModule(state.Block)
+
+	var res ChainHashResponse
+	req := ChainFinalizedHeadRequest{0, 0}
+	err := svc.GetFinalizedHeadByRound(nil, &req, &res)
+	require.NoError(t, err)
+
+	_, _, genesisHeader := genesis.NewTestGenesisWithTrieAndHeader(t)
+	expected := genesisHeader.Hash()
+	require.Equal(t, common.BytesToHex(expected[:]), res)
+
+	digest := types.NewDigest()
+	prd, err := types.NewBabeSecondaryPlainPreDigest(0, 1).ToPreRuntimeDigest()
+	require.NoError(t, err)
+	err = digest.Add(*prd)
+	require.NoError(t, err)
+	header := &types.Header{
+		ParentHash: genesisHeader.Hash(),
+		Number:     big.NewInt(1),
+		Digest:     digest,
+	}
+	err = state.Block.AddBlock(&types.Block{
+		Header: *header,
+		Body:   types.Body{},
+	})
+	require.NoError(t, err)
+
+	testhash := header.Hash()
+	err = state.Block.SetFinalisedHash(testhash, 77, 1)
+	require.NoError(t, err)
+
+	req = ChainFinalizedHeadRequest{77, 1}
+	err = svc.GetFinalizedHeadByRound(nil, &req, &res)
+	require.NoError(t, err)
+	require.Equal(t, common.BytesToHex(testhash[:]), res)
+}
+
+func newTestStateService(t *testing.T) *state.Service {
+	testDatadirPath, err := ioutil.TempDir("/tmp", "test-datadir-*")
+	require.NoError(t, err)
+
+	config := state.Config{
+		Path:     testDatadirPath,
+		LogLevel: log.Info,
+	}
+	stateSrvc := state.NewService(config)
+	stateSrvc.UseMemDB()
+
+	gen, genTrie, genesisHeader := genesis.NewTestGenesisWithTrieAndHeader(t)
+
+	err = stateSrvc.Initialise(gen, genesisHeader, genTrie)
+	require.NoError(t, err)
+
+	err = stateSrvc.Start()
+	require.NoError(t, err)
+
+	rtCfg := &wasmer.Config{}
+
+	rtCfg.Storage, err = rtstorage.NewTrieState(genTrie)
+	require.NoError(t, err)
+
+	if stateSrvc != nil {
+		rtCfg.NodeStorage.BaseDB = stateSrvc.Base
+	} else {
+		rtCfg.NodeStorage.BaseDB, err = utils.SetupDatabase(filepath.Join(testDatadirPath, "offline_storage"), false)
+		require.NoError(t, err)
+	}
+
+	rt, err := wasmer.NewRuntimeFromGenesis(rtCfg)
+	require.NoError(t, err)
+
+	loadTestBlocks(t, genesisHeader.Hash(), stateSrvc.Block, rt)
+
+	t.Cleanup(func() {
```

Comment:
Maybe move that up as soon as `stateSrvc` is defined? In case the test fails with a `require` instruction